app/models/concerns/profile_fields.rb

```diff
@@ -47,5 +65,23 @@ def bm_versions
         upstream: false
       ).joins(:benchmark).pluck('benchmarks.version')
     end
+
+    def conflicting_relationships(rules, rule_groups)
+      conflicting_relationships = {}
+      RuleGroupRelationship.with_conflicts(rules).or(RuleGroupRelationship.with_conflicts(rule_groups)).each do |rgr|
```

Comment:
Iteration is always a red flag :red_circle: :flags: calling an `each` and then manipulating data (and not using map or reduce) in its body is an even bigger one. I think you can do this operation in SQL but at least use an [`each_with_object`](https://womanonrails.com/each-with-object) please.

The in-SQL might bee too complex, but you can look into the `ARRAY_AGG` and `JSON_AGG` queries. We can create a follow-up ticket where we optimize this in SQL and so eliminate these helper methods and move them purely into scopes.

Reply:
I think this probably could be done in SQL but if I used the associations that I added a comment about, I don't think these helper methods will be needed at all. Looking into the algorithmic complexity of the association solution now.